session/manager.go

```diff
@@ -162,6 +164,42 @@ func (sm *Manager) GetSessionHistory() map[string]*History {
 	return historyCopy
 }
 
+var ErrNotFound = errors.New("session not found")
+
+func (sm *Manager) GetSessionFromHistory(sessionID string) (*History, error) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+	for id, history := range sm.history {
+		if id == sessionID {
+			return history, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+func (sm *Manager) CancelSession(sessionID, reason string) error {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+	var session *Session
+	for id, s := range sm.sessions {
+		if id == sessionID {
+			session = &s.Session
```

Comment:
no point in continuing once found, I would add a `break` after this.

Reply:
also, why am I looping at all here? It's a hash map